src/librustc_resolve/macros.rs

```diff
@@ -475,6 +475,10 @@ impl<'a, 'cl> Resolver<'a, 'cl> {
             return def;
         }
 
+        if kind == MacroKind::Attr && *&path[0].as_str() == "test" {
```

Comment:
cc @petrochenkov, do you know how this might be best implemented? 

The gist of the change here is that `#[test]` is becoming a procedural macro attribute, but this is in practice conflicting with a lot of `macro_rules!` macros called `test`. I believe the intention here is that `test` as an *attribute* is always resolved to the macro attribute, but otherwise it goes through normal resolution rules.

@djrenren I think what we probably want, long term, here is to have a forward-compatibility warning issued. Crates which require this overriding behavior of "`#[test]` trumps `test!`" will want to get a warning that eventually we'd like to remove this hack in the future, and the macro should be renamed.

Reply:
I'm all for it :+1: